store/root/store.go

```diff
@@ -0,0 +1,286 @@
+}
+
+// GetProof delegates the GetProof to the store's underlying SC backend.
+func (s *Store) GetProof(_ string, version uint64, key []byte) (*ics23.CommitmentProof, error) {
+	return s.stateCommitment.GetProof(version, key)
+}
+
+// LoadVersion loads a specific version returning an error upon failure.
+func (s *Store) LoadVersion(v uint64) (err error) {
+	return s.loadVersion(v, nil)
+}
+
+// GetKVStore returns the store's root KVStore. Any writes to this store without
+// branching will be committed to SC and SS upon Commit(). Branching will create
+// a branched KVStore that allow writes to be discarded and propagated to the
+// root KVStore using Write().
+func (s *Store) GetKVStore(_ string) store.KVStore {
+	return s.rootKVStore
+}
+
+func (s *Store) loadVersion(v uint64, upgrades any) error {
+	s.logger.Debug("loading version", "version", v)
+
+	if err := s.stateCommitment.LoadVersion(v); err != nil {
+		return fmt.Errorf("failed to load SS version %d: %w", v, err)
+	}
+
+	// TODO: Complete this method to handle upgrades. See legacy RMS loadVersion()
+	// for reference.
+	//
+	// Ref: https://github.com/cosmos/cosmos-sdk/issues/17314
+
+	return nil
+}
+
+func (s *Store) WorkingHash() []byte {
+	storeInfos := []store.StoreInfo{
+		{
+			Name: defaultStoreKey,
+			CommitID: store.CommitID{
+				Hash: s.stateCommitment.WorkingHash(),
+			},
+		},
+	}
+
+	return store.CommitInfo{StoreInfos: storeInfos}.Hash()
+}
+
+func (s *Store) SetCommitHeader(h store.CommitHeader) {
+	s.commitHeader = h
+}
+
+// Commit commits all state changes to the underlying SS backend and all SC stores.
+// Note, writes to the SS backend are retrieved from the SC memory listeners and
+// are committed in a single batch synchronously. All writes to each SC are also
+// committed synchronously, however, they are NOT atomic. A byte slice is returned
+// reflecting the Merkle root hash of all committed SC stores.
+func (s *Store) Commit() ([]byte, error) {
+	var previousHeight, version uint64
+	if s.lastCommitInfo.GetVersion() == 0 && s.initialVersion > 1 {
+		// This case means that no commit has been made in the store, we
+		// start from initialVersion.
+		version = s.initialVersion
+	} else {
+		// This case can means two things:
+		//
+		// 1. There was already a previous commit in the store, in which case we
+		// 		increment the version from there.
+		// 2. There was no previous commit, and initial version was not set, in which
+		// 		case we start at version 1.
+		previousHeight = s.lastCommitInfo.GetVersion()
+		version = previousHeight + 1
+	}
+
+	if s.commitHeader.GetHeight() != version {
+		s.logger.Debug("commit header and version mismatch", "header_height", s.commitHeader.GetHeight(), "version", version)
+	}
+
+	changeSet := s.rootKVStore.GetChangeSet()
+
+	// commit SS
+	if err := s.commitSS(version, changeSet); err != nil {
+		return nil, fmt.Errorf("failed to commit SS: %w", err)
+	}
+
+	// commit SC
+	commitInfo, err := s.commitSC(version, changeSet)
```

Comment:
it seems like the finalizeBlock step is missing, the current BaseApp flow looks like WriteBatch->WorkingHash->Commit. Maybe we need to separate the writing from Commit

Reply:
https://github.com/cosmos/cosmos-sdk/pull/17577/commits/451d994bfb5c10203e15c7c1e1eb1d55f370c338